File: Algoritmos/Listas/Doubly_Linked_List.cpp
/*
Implementation of a doubly linked list

iterating the list

for (node<type> *n = l.begin(); n != l.end(); n = n->next())
{
    ...
}

*/

#include <iostream>
#include <memory>
#include <stdexcept>

template <typename type>
class list;

template <typename type>
class node
{
    private:
        node<type> *next_node;
        node<type> *prev_node;
        list<type> *origin;
    
        friend class list<type>;

    public:
        type elem;

        node(type elem, node *next, node *prev, list<type> *origin) : elem(elem), next_node(next), prev_node(prev), origin(origin) {}

        // Returns the next node of n
        node<type>* next() const
        {
            if (this == origin->tail)
                throw std::out_of_range("the node is out of the list range");

            return next_node;
        }

        // Returns the previous node of n
        node<type>* prev() const
        {
            if (this == origin->head)
                throw std::out_of_range("the node is out of the list range");

            return prev_node;
        }
};

template <typename type>
class list
{
    private:
        size_t size;
        node<type> *head;
        node<type> *tail;

        friend class node<type>;

    public:
        list()
        {
            size = 0;

            head = new node<type>({}, nullptr, nullptr, this);
            tail = new node<type>({}, nullptr, nullptr, this);

            head->next_node = tail;
            head->prev_node = nullptr;

            tail->next_node = nullptr;
            tail->prev_node = head;
        }
        list(const list<type> &other)
        {
            size = 0;

            head = new node<type>({}, nullptr, nullptr, this);
            tail = new node<type>({}, nullptr, nullptr, this);

            head->next_node = tail;
            head->prev_node = nullptr;

            tail->next_node = nullptr;
            tail->prev_node = head;

            push_back_list(other);
        }
        ~list()
        {
            clear();
            delete head;
            delete tail;
        }

        // Returns the size of the list
        int get_size() const { return size; }

        // Returns true if the list is empty, and returns false if not
        bool empty() const { return size == 0; }

        // Returns true if the node is part of the list, and returns false if not
        bool in_list(const node<type> *n) const { return n->origin == this; }

        // Returns the first element of the list
        node<type> *begin() const { return head->next_node; }

        // Returns the tail (after the last element) of the list
        node<type> *end() const { return tail; }

        // Insert elem to the end of the list
        void push_back(type elem)
        {
            node<type> *new_node = new node<type>(elem, tail, tail->prev_node, this);

            tail->prev_node->next_node = new_node;
            tail->prev_node = new_node;

            size++;
        }

        // Insert elem to the begining of the list
        void push_front(type elem)
        {
            node<type> *new_node = new node<type>(elem, head->next_node, head, this);

            head->next_node = new_node;
            new_node->next_node->prev_node = new_node;

            size++;
        }

        // inserts the node to the end of the list
        // the node must not be part of any list, including this
        void push_back_node(node<type> *n)
        {
            if (n == nullptr || n == head || n == tail || n->origin != nullptr)
                throw std::invalid_argument("invalid node passed as argument to push_back_node()");
            
            n->origin = this;

            tail->prev_node->next_node = n;
            tail->prev_node = n;

            size++;
        }

        // inserts the node to the begining of the list
        // the node must not be part of any list, including this
        void push_front_node(node<type> *n)
        {
            if (n == nullptr || n == head || n == tail || n->origin != nullptr)
                throw std::invalid_argument("invalid node passed as argument to push_back_node()");
            
            n->origin = this;

            head->next_node = n;
            n->next_node->prev_node = n;

            size++;
        }
        
        // Insert copies of the other list elements to the end of this list
        void push_back_list(const list<type> &other)
        {
            if (other.empty())
                return;

            node<type> *n = other.begin();
            while (n != other.end())
            {
                push_back(n->elem);
                n = n->next();
            }
        }

        // Insert copies of the other list elements to the begining of this list
        void push_front_list(const list &other)
        {
            if (other.empty())
                return;
            
            node<type> *n = other.end()->prev_node;
            while (n != other.begin()->prev_node)
            {
                push_front(n->elem);
                n = n->prev();
            }
        }

        // Insert the elements of the other list elements to the end of this list
        // and empties the other list
        void push_back_transfer_list(list<type> &other)
        {
            if (other.empty())
                return;

            // Update the origin of the nodes being transferred
            node<type> *n = other.begin();
            while (n != other.end())
            {
                n->origin = this;
                n = n->next_node;
            }

            tail->prev_node->next_node = other.head->next_node;
            other.head->next_node->prev_node = tail->prev_node;

            tail->prev_node = other.tail->prev_node;
            other.tail->prev_node->next_node = tail;

            size += other.size;
            
            // Clear other list
            other.head->next_node = other.tail;
            other.tail->prev_node = other.head;
            other.size = 0;        
        }

        // Insert the elements of the other list elements to the begining of this list
        // and empties the other list
        void push_front_transfer_list(list<type> &other)
        {
            if (other.empty())
                return;
            
            // Update the origin of the nodes being transferred
            node<type> *n = other.begin();
            while (n != other.end())
            {
                n->origin = this;
                n = n->next_node;
            }

            head->next_node->prev_node = other.tail->prev_node;
            other.tail->prev_node->next_node = head->next_node;

            head->next_node = other.head->next_node;
            other.head->next_node->prev_node = head;

            size += other.size;
            
            // Clear other list
            other.head->next_node = other.tail;
            other.tail->prev_node = other.head;
            other.size = 0;        
        }


        // Removes the last element
        type pop_back()
        {
            if (size == 0)
                throw std::runtime_error("Cannot pop from an empty list.");
            
            node<type> *temp = tail->prev_node;
            type elem = temp->elem;

            tail->prev_node = tail->prev_node->prev_node;
            tail->prev_node->next_node = tail;
            size--;
            
            delete temp;
            return elem;
        }

        // Removes the first element
        type pop_front()
        {
            if (size == 0)
                throw std::runtime_error("Cannot pop from an empty list.");
            
            node<type> *temp = head->next_node;
            type elem = temp->elem;

            head->next_node = head->next_node->next_node;
            head->next_node->prev_node = head;
            size--;
            
            delete temp;
            return elem;
        }

        // Removes the node n if it belongs to this list 
        node<type> pop_node(node<type> *n)
        {
            if (n == nullptr)
                throw std::invalid_argument("nullptr passed as argument to pop_node()");
            else if (n == head || n == tail)
                throw std::out_of_range("The node is out of range");
            else if (!in_list(n))
                throw std::invalid_argument("The node does not belong to this list");
            
            n->prev_node->next_node = n->next_node;
            n->next_node->prev_node = n->prev_node;
            n->origin = nullptr;
            size--;
            
            return n;
        }

        // Removes all the occurences of elem
        void remove(type elem)
        {
            for (node<type> *i = begin(); i != end();)
            {
                if (i->elem == elem)
                {
                    node<type> *temp = i;
                    
                    i->prev_node->next_node = i->next_node;
                    i->next_node->prev_node = i->prev_node;

                    i = next(i);
                    delete temp;
                    size--;
                }
                else
                    i = next(i);
            }
        }

        // Removes all the elements
        void clear()
        {
            while (size > 0)
            {
                pop_front();
            }
        }


        void print_list() const
        {
            node<type> *curr = begin();
            
            while (curr != end())
            {
                std::cout << curr->elem << ' ';
                curr = curr->next_node;
            }
            std::cout << '\n';
        }
};

File: Algoritmos/Listas/doubly_linked_list.go
/*

This Package implements a doubly linked list

To iterate over a list (where l is a *List[T]):

for n := l.Begin(); n != l.End().Next(); n = n.Next() {
	// do something with n.Elem
}

*/

package list

import (
	"errors"
)

type node[T any] struct  {
	Elem T
	next *node[T]
	prev *node[T]
	list *List[T]
}

// Returns a new and initialized node
func NewNode[T any](elem T, next *node[T], prev *node[T], list *List[T]) *node[T] {
	return &node[T] {elem, next, prev, list}
}

// Returns the next node or nil
func (n *node[T]) Next() *node[T] {
	if n == nil {
		return nil
	}
	return n.next
}

// Returns the previous node or nil
func (n *node[T]) Prev() *node[T] {
	if n == nil {
		return nil
	}
	return n.prev
}


type List[T any] struct {
	size uint
	head *node[T]
	tail *node[T]
}

// Returns a new list
func NewList[T any]() *List[T] {
	l := &List[T] {
		size: 0,
		head: nil,
		tail: nil,
	}

	l.head = NewNode(*new(T), nil, nil, l)
	l.tail = NewNode(*new(T), nil, nil, l)
	
	l.head.next = l.tail
	l.tail.prev = l.head

	return l
}

// Checks if the node n is in the list
func (l *List[T]) InList(n *node[T]) bool {
	return n.list == l
}

// Returns the first node of the list
func (l *List[T]) Begin() *node[T] {
	return l.head.next
}

// Returns the last node of the list
func (l *List[T]) End() *node[T] {
	return l.tail.prev
}

// Returns true if the list is empty, otherwise, it will return false
func (l *List[T]) Empty() bool {
	return l.head.next == l.tail 
}


// Inserts the element in the end of the list
func (l *List[T]) PushBack(elem T) {
	n := NewNode(elem, l.tail, l.tail.prev, l)

	n.prev.next = n
	l.tail.prev = n
	l.size++
}

// Inserts the element in the begining of the list
func (l *List[T]) PushFront(elem T) {
	n := NewNode(elem, l.head.next, l.head, l)

	n.next.prev = n
	l.head.next = n

	l.size++
}

// Inserts a copy of another list at the back of list l.
func (l *List[T]) PushBackList(other *List[T]) error {
	if other == nil {
		return errors.New("other list is nil")
	} else if other.size == 0 {
		return nil
	}
	
	var n *node[T] = other.Begin()
	for n != other.tail {
		l.PushBack(n.Elem)
		n = n.next
	}

	l.size += other.size
	return nil
}

// Inserts a copy of another list at the front of list l.
func (l *List[T]) PushFrontList(other *List[T]) error {
	if other == nil {
		return errors.New("other list is nil")
	} else if other.size == 0 {
		return nil
	}
	
	var n *node[T] = other.End()
	for n != other.head {
		l.PushFront(n.Elem)
		n = n.prev
	}

	l.size += other.size
	return nil
}

// Inserts a another list at the back of list l
// The other list becomes empty
func (l *List[T]) PushBack_ClearList(other *List[T]) error {
	if other == nil {
		return errors.New("other list is nil")
	} else if other.size == 0 {
		return nil
	}
	
	l.tail.prev.next = other.head.next
	other.head.next.prev = l.tail.prev

	l.tail.prev = other.tail.prev
	other.tail.prev.next = l.tail

	// Update the list reference for nodes
	for n := other.head.next; n != other.tail; n = n.next {
		n.list = l
	}

	// Empty the other list
	other.head.next = other.tail
	other.tail.prev = other.head
	other.size = 0

	l.size += other.size
	return nil
}

// Inserts a another list at the front of list l
// The other list becomes empty
func (l *List[T]) PushFront_ClearList(other *List[T]) error {
	if other == nil {
		return errors.New("other list is nil")
	} else if other.size == 0 {
		return nil
	}

	l.head.next.prev = other.tail.prev
	other.tail.prev.next = l.head.next

	l.head.next = other.head.next
	other.head.next.prev = l.head

	// Update the list reference for nodes
	for n := other.head.next; n != other.tail; n = n.next {
		n.list = l
	}

	// Empty the other list
	other.head.next = other.tail
	other.tail.prev = other.head
	other.size = 0

	l.size += other.size
	return nil
}

// Inserts the element after the specified node
func (l *List[T]) InsertAfter(elem T, n *node[T]) error {
	if n == nil {
		return errors.New("node is nil")
	} else if n == l.tail {
		return errors.New("insertion is out of list range")
	} else if !l.InList(n) {
		return errors.New("the node is not on the list")
	}
	
	new_node := NewNode(elem, n.next, n, l)
	n.next.prev = new_node
	n.next = new_node

	return nil
}

// Inserts the element before the specified node
func (l *List[T]) InsertBefore(elem T, n *node[T]) error {
	if n == nil {
		return errors.New("node is nil")
	} else if n == l.head {
		return errors.New("insertion is out of list range")
	} else if !l.InList(n) {
		return errors.New("the node is not on the list")
	}
	
	new_node := NewNode(elem, n, n.prev, l)
	n.prev.next = new_node
	n.prev = new_node

	return nil
}


// Removes the last element of the list
func (l *List[T]) PopBack() (T, error) {
	if l.Empty() {
		return *new(T), errors.New("list is empty")
	}
	
	n := l.tail.prev

	l.tail.prev = l.tail.prev.prev
	l.tail.prev.next = l.tail
	l.size--

	n.next = nil
	n.prev = nil

	return n.Elem, nil
}

// Removes the first node of the list
func (l *List[T]) PopFront() (T, error) {
    if l.Empty() {
        return *new(T), errors.New("list is empty")
    }

    n := l.head.next

    l.head.next = l.head.next.next
    l.head.next.prev = l.head
	l.size--

	n.next = nil
	n.prev = nil

    return n.Elem, nil
}

// Removes the specified node of the list
func (l *List[T]) PopNode(n *node[T]) (*node[T], error) {
	if n == nil {
		return nil, errors.New("node is nil")
	} else if n == l.head || n == l.tail {
		return nil, errors.New("node is out of range")
	} else if !l.InList(n) {
		return nil, nil
	}

	n.prev.next = n.next
	n.next.prev = n.prev
	l.size--

	n.next = nil
	n.prev = nil
	n.list = nil

	return n, nil
}

// Removes all ocurrences of elem using the provided comparison function
func (l *List[T]) Remove(elem T, compFunc func(T, T) bool) {
	for n := l.Begin(); n != l.End().next; {
		if compFunc(n.Elem, elem) {
			n = n.next
			l.PopNode(n.prev)
		} else {
			n = n.next
		}
	}
}

// Removes all the elements of the list
func (l *List[T]) Clear() {
	l.head.next = l.tail
	l.tail.prev = l.head
	l.size = 0
}

// Moves the n node to after mark node
func (l *List[T]) MoveAfter(n *node[T], mark *node[T]) {
	if n == nil || mark == nil || n == mark || n == l.head || n == l.tail || 
		mark == l.tail || !l.InList(n) || !l.InList(mark) {
		return
	}

	n.prev.next = n.next
	n.next.prev = mark.prev
	
	n.next = mark.next
	n.prev = mark

	mark.next.prev = n
	mark.next = n
}

// Moves the n node to before mark node
func (l *List[T]) MoveBefore(n *node[T], mark *node[T]) {
	if n == nil || mark == nil || n == mark || n == l.head || n == l.tail ||
	   mark == l.head || !l.InList(n) || !l.InList(mark) {
		return
	}

	n.prev.next = n.next
	n.next.prev = mark.prev

	n.next = mark
	n.prev = mark.prev

	mark.prev.next = n
	mark.prev = n
}

// Moves the node n of the list to the back 
func (l *List[T]) MoveToBack(n *node[T]) {
	if n == nil || n == l.head || n == l.tail || !l.InList(n) {
		return
	}

	n.next.prev = n.prev
	n.prev.next = n.next

	n.next = l.head.next
	n.prev = l.head

	l.head.next = n
	n.next.prev = n
}

// Moves the node n of the list to the front 
func (l *List[T]) MoveToFront(n *node[T]) {
	if n == nil || n == l.head || n == l.tail || !l.InList(n) {
		return
	}

	n.next.prev = n.prev
	n.prev.next = n.next

	n.next = l.tail
	n.prev = l.tail.prev

	l.tail.prev.next = n
	l.tail.prev = n
}

// Reverses the order of elements in the list
func (l *List[T]) Reverse() {
	if l.size <= 1 {
		return
	}

	n := l.head.next
	l.head.next = l.tail.prev
	l.tail.prev = n

	for ; n != l.tail; n = n.prev {
		n.next, n.prev = n.prev, n.next
	}
}
